### PR TITLE
Modify index bound value

### PR DESCRIPTION
**Pull request by cangfengzhs**
#### What type of PR is this?
- [x] bug
- [ ] feature
- [ ] enhancement

#### What does this PR do?
close issue 2745

#### Which issue(s)/PR(s) this PR relates to?
Fix the problem of index query interval error
  
#### Checklist：
- [ ] Documentation affected （Please add the label if documentation needs to be modified.)
- [ ] Incompatible （If it is incompatible, please describe it and add corresponding label.）
- [ ] Need to cherry-pick （If need to cherry-pick to some branches, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

#### Release notes：
Please confirm whether to reflect in release notes and how to describe:
>                                                                 `

### CODE REVIEW

File: src/graph/optimizer/rule/IndexScanRule.cpp [hunk 1]
```diff
@@ -199,46 +199,101 @@ Status IndexScanRule::appendIQCtx(const IndexItem& index, IndexQueryCtx& iqctx)
...
-      begin = OptimizerUtils::normalizeValue(col, item.value_);
+      begin = {item.value_, true};
       break;
     }
     // because only type for bool is true/false, which can not satisify [start,
     // end)
     if (col.get_type().get_type() == nebula::cpp2::PropertyType::BOOL) {
       return Status::SemanticError("Range scan for bool type is illegal");
     }
-    NG_RETURN_IF_ERROR(OptimizerUtils::boundValue(item.relOP_, item.value_, col, begin, end));
```
Comment (czpmango) [inline src/graph/optimizer/rule/IndexScanRule.cpp:224]: Delete all boundValue related functions.

File: src/graph/optimizer/OptimizerUtils.cpp [hunk 2]
```diff
@@ -949,5 +949,15 @@ void OptimizerUtils::copyIndexScanData(const nebula::graph::IndexScan* from,
...
+Status OptimizerUtils::compareAndSwapBound(std::pair<Value, bool>& a, std::pair<Value, bool>& b) {
+  if (a.first > b.first) {
+    std::swap(a, b);
+  } else if (a.first < b.first) {  // do nothing
+  } else if (a.second > b.second) {
+    std::swap(a, b);
+  }
+  return Status::OK();
+}
+
```
Comment (czpmango) [inline src/graph/optimizer/OptimizerUtils.cpp:961]: This is so weird as a utility function.

File: tests/tck/features/optimizer/IndexScanRule.feature [hunk 3]
```diff
@@ -20,25 +20,25 @@ Feature: Match index selection
...
       MATCH (v:player)
       WHERE
-        v.name<="Aron Baynes"
-        or v.name>"Yao Ming"
-        or v.name=="Kobe Bryant"
-        or v.age>40
+      v.name<="Aron Baynes"
+      or v.name>"Yao Ming"
+      or v.name=="Kobe Bryant"
+      or v.age>40
```
Comment (czpmango) [inline tests/tck/features/optimizer/IndexScanRule.feature:41]: Is this a better format?

File: src/graph/optimizer/rule/IndexScanRule.cpp [hunk 4]
```diff
@@ -199,46 +199,101 @@ Status IndexScanRule::appendIQCtx(const IndexItem& index, IndexQueryCtx& iqctx)
...
+        include = true;
+        [[fallthrough]];
+      case Expression::Kind::kRelLT: {
+        if (end.first.empty()) {
+          end.first = item.value_;
+          end.second = include;
+        } else {
+          auto tmp = std::make_pair(item.value_, include);
+          OptimizerUtils::compareAndSwapBound(end, tmp);
+        }
```
Comment (czpmango) [inline src/graph/optimizer/rule/IndexScanRule.cpp:264]: What does this code mean？

File: src/graph/optimizer/rule/IndexScanRule.cpp (hunk 1 above)
Comment (cangfengzhs) [reply src/graph/optimizer/rule/IndexScanRule.cpp:224]: There are some other functions that still use them, such as `mergeRangeColumnHints`. The logic of these functions is very complicated, so I didn't modify them.

Comment (czpmango) [reply src/graph/optimizer/rule/IndexScanRule.cpp:224]: If this function is called, isn't the value it changes illegal？
If we add an interval open and close flag, we shouldn't do boundValue anymore.

File: tests/tck/features/optimizer/IndexScanRule.feature [hunk 5]
```diff
@@ -20,25 +20,25 @@ Feature: Match index selection
...
-      | 6  | Project     | 5            |                                                                                                                        |
-      | 5  | Filter      | 15           |                                                                                                                        |
-      | 15 | GetVertices | 11           |                                                                                                                        |
-      | 11 | IndexScan   | 0            | {"indexCtx": {"columnHints":{"scanType":"RANGE","column":"name","beginValue":"\"Tim Duncan","endValue":"\"Yao Ming"}}} |
-      | 0  | Start       |              |                                                                                                                        |
+      | id | name        | dependencies | operator info                                                                                                              |
+      | 10 | Project     | 13           |                                                                                                                            |
```
Comment (czpmango) [inline tests/tck/features/optimizer/IndexScanRule.feature:24]: Add the following case and profile the `begin_value` and `end_value` in `ColumnHints`
```
profile match (v:player) where v.age>30 and v.age<=40 return v
```

File: tests/tck/features/optimizer/IndexScanRule.feature (hunk 3 above)
Comment (cangfengzhs) [reply tests/tck/features/optimizer/IndexScanRule.feature:41]: This is automatically formatted by vscode. It was not in this format before, but lint would report an error, so I formatted it with vscode.

File: src/graph/optimizer/rule/IndexScanRule.cpp (hunk 1 above)
Comment (cangfengzhs) [reply src/graph/optimizer/rule/IndexScanRule.cpp:224]: OK, I have deleted `boundValue` just now, but the logic of `selectIndex` is very complicated, and there are a lot of code repetitions. I am more inclined to reorganize this part of the logic when you reconstruct the logic of selecting indexes.

File: tests/tck/features/optimizer/IndexScanRule.feature (hunk 5 above)
Comment (cangfengzhs) [reply tests/tck/features/optimizer/IndexScanRule.feature]: OK

File: src/graph/optimizer/OptimizerUtils.cpp (hunk 2 above)
Comment (cangfengzhs) [reply src/graph/optimizer/OptimizerUtils.cpp:425]: I add some comment to this function.

File: src/graph/optimizer/rule/IndexScanRule.cpp (hunk 1 above)
Comment (czpmango) [reply src/graph/optimizer/rule/IndexScanRule.cpp:224]: Ok.

File: src/graph/optimizer/rule/IndexScanRule.cpp (hunk 4 above)
Comment (cangfengzhs) [reply src/graph/optimizer/rule/IndexScanRule.cpp:264]: assign min(end,tmp) to end

File: src/graph/optimizer/OptimizerUtils.cpp [hunk 6]
```diff
@@ -724,7 +208,15 @@ bool mergeRangeColumnHints(const ColumnDef& field,
       }
     }
   }
-  return !(*begin >= *end);
+  bool ret = true;
+  if (begin->first > end->first) {
+    ret = false;
+  } else if (begin->first == end->first) {
+    if (begin->second == false && end->second == false) {
```
Comment (critical27) [inline src/graph/optimizer/OptimizerUtils.cpp:215]: What will happen we pass interval of (x, x)? Report error?

File: src/graph/optimizer/OptimizerUtils.cpp [hunk 7]
```diff
@@ -748,34 +240,33 @@ bool getIndexColumnHintInExpr(const ColumnDef& field,
...
-  } else if (newBegin == end) {
-    // end == neBegin == begin + 1
-    h.hint.set_scan_type(storage::cpp2::ScanType::PREFIX);
-    h.hint.set_begin_value(std::move(begin));
-    h.score = IndexScore::kPrefix;
+  } else if (begin.first == end.first) {
+    if (begin.second == false && end.second == false) {
+      return false;
+    } else {
+      h.hint.set_scan_type(storage::cpp2::ScanType::RANGE);
```
Comment (critical27) [inline src/graph/optimizer/OptimizerUtils.cpp:260]: All PREFIX has convert to [value, value]?

File: src/graph/optimizer/OptimizerUtils.cpp [hunk 8]
```diff
@@ -748,34 +240,33 @@ bool getIndexColumnHintInExpr(const ColumnDef& field,
...
+    h.hint.set_include_begin(begin.second);
+    h.hint.set_include_end(end.second);
     h.score = IndexScore::kRange;
-  } else if (newBegin == end) {
-    // end == neBegin == begin + 1
-    h.hint.set_scan_type(storage::cpp2::ScanType::PREFIX);
-    h.hint.set_begin_value(std::move(begin));
-    h.score = IndexScore::kPrefix;
+  } else if (begin.first == end.first) {
+    if (begin.second == false && end.second == false) {
```
Comment (critical27) [inline src/graph/optimizer/OptimizerUtils.cpp:257]: Why we don't need to return false for case (x, x], [x, x)?
> (x, x) has return false in line 244

File: src/graph/optimizer/OptimizerUtils.cpp (hunk 6 above)
Comment (cangfengzhs) [reply src/graph/optimizer/OptimizerUtils.cpp:215]: No errors will be reported and will be executed normally. But no data will be returned.

File: src/graph/optimizer/OptimizerUtils.cpp (hunk 7 above)
Comment (cangfengzhs) [reply src/graph/optimizer/OptimizerUtils.cpp:260]: No. This is only used to calculate whether the union of multiple IndexItems is an empty set (begin>end)

File: src/graph/optimizer/OptimizerUtils.cpp (hunk 8 above)
Comment (cangfengzhs) [reply src/graph/optimizer/OptimizerUtils.cpp:257]: This is a good suggestion. I was not sure if (x,x] or [x,x) represented the empty set.

File: src/graph/optimizer/OptimizerUtils.cpp (hunk 6 above)
Comment (critical27) [reply src/graph/optimizer/OptimizerUtils.cpp]: 👌

**Review by critical27 — APPROVED** [review]

**Review by czpmango — APPROVED** [review]

**Pull request merged**